// mongoose-data-seed/dist/bin/run/options.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getOptions = exports.optionDefinitions = undefined;

var _commandLineArgs2 = require('command-line-args');

var _commandLineArgs3 = _interopRequireDefault(_commandLineArgs2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var optionDefinitions = exports.optionDefinitions = [{
  name: 'seeders',
  alias: 's',
  type: String,
  multiple: true,
  defaultValue: [],
  defaultOption: true,
  typeLabel: 'seeder ...',
  description: 'Seeders names to run'
}, {
  name: 'dropdb',
  alias: 'd',
  type: Boolean,
  defaultValue: false,
  description: 'Drop the database before seeding'
}, {
  name: 'help',
  alias: 'h',
  type: Boolean,
  defaultValue: false,
  description: 'Show usage guide'
}];

var getOptions = exports.getOptions = function getOptions(argv) {
  var _commandLineArgs = (0, _commandLineArgs3.default)(optionDefinitions, { argv: argv }),
      selectedSeeders = _commandLineArgs.seeders,
      dropDatabase = _commandLineArgs.dropdb,
      helpWanted = _commandLineArgs.help;

  return { selectedSeeders: selectedSeeders, dropDatabase: dropDatabase, helpWanted: helpWanted };
};

exports.default = getOptions;
