// mongoose-data-seed/dist/lib/config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _findRoot = require('find-root');

var _findRoot2 = _interopRequireDefault(_findRoot);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var workingDir = process.cwd();
var projectRoot = (0, _findRoot2.default)(workingDir);

var userGeneratorConfigFilename = 'md-seed-generator.json';
var userGeneratorConfigFilepath = _path2.default.join(projectRoot, userGeneratorConfigFilename);
var userGeneratorConfigExists = _fs2.default.existsSync(userGeneratorConfigFilepath);

var _ref = userGeneratorConfigExists ? require(userGeneratorConfigFilepath) : {},
    _ref$es = _ref.es6,
    useEs6Generator = _ref$es === undefined ? true : _ref$es,
    _ref$seedersFolder = _ref.seedersFolder,
    userSeedersFolderName = _ref$seedersFolder === undefined ? 'seeders' : _ref$seedersFolder;

var userSeedersFolderPath = _path2.default.join(projectRoot, userSeedersFolderName);

var userConfigFilename = 'md-seed-config.js';
var userConfigFilepath = _path2.default.join(projectRoot, userConfigFilename);
var userConfigExists = _fs2.default.existsSync(userConfigFilepath);

var userConfig = userConfigExists ? require(userConfigFilepath) : null;

var config = {
  workingDir: workingDir,
  projectRoot: projectRoot,
  useEs6Generator: useEs6Generator,
  userGeneratorConfigFilename: userGeneratorConfigFilename,
  userGeneratorConfigFilepath: userGeneratorConfigFilepath,
  userGeneratorConfigExists: userGeneratorConfigExists,
  userConfigFilename: userConfigFilename,
  userConfigFilepath: userConfigFilepath,
  userSeedersFolderName: userSeedersFolderName,
  userSeedersFolderPath: userSeedersFolderPath,
  userConfigExists: userConfigExists,
  userConfig: userConfig
};

exports.default = config;
