// mongoose-data-seed/dist/bin/init/options.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getOptions = exports.optionDefinitions = undefined;

var _commandLineArgs2 = require('command-line-args');

var _commandLineArgs3 = _interopRequireDefault(_commandLineArgs2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var optionDefinitions = exports.optionDefinitions = [{
  name: 'es6',
  type: Boolean,
  description: 'Use es6 syntax, require babel'
}, {
  name: 'seedersFolder',
  alias: 'f',
  type: String,
  description: 'Seeders folder name'
}, {
  name: 'help',
  alias: 'h',
  type: Boolean,
  defaultValue: false,
  description: 'Show usage guide'
}];

var getOptions = exports.getOptions = function getOptions(argv) {
  var _commandLineArgs = (0, _commandLineArgs3.default)(optionDefinitions, { argv: argv }),
      es6 = _commandLineArgs.es6,
      seedersFolder = _commandLineArgs.seedersFolder,
      helpWanted = _commandLineArgs.help;

  return { es6: es6, seedersFolder: seedersFolder, helpWanted: helpWanted };
};

exports.default = getOptions;
